php/php.project/src/org/netbeans/modules/php/project/ComputeTestMethodAnnotations.java

```diff
@@ -0,0 +1,193 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.php.project;
+
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import javax.swing.text.BadLocationException;
+import javax.swing.text.Document;
+import javax.swing.text.JTextComponent;
+import org.netbeans.api.editor.EditorRegistry;
+import org.netbeans.modules.editor.NbEditorUtilities;
+import org.netbeans.modules.gsf.testrunner.ui.api.TestMethodController;
+import org.netbeans.modules.gsf.testrunner.ui.api.TestMethodController.TestMethod;
+import org.netbeans.modules.php.api.editor.EditorSupport;
+import org.netbeans.modules.php.api.editor.PhpClass;
+import org.netbeans.modules.php.api.editor.PhpType;
+import org.netbeans.modules.php.api.phpmodule.PhpModule;
+import org.netbeans.modules.php.api.util.FileUtils;
+import org.netbeans.modules.php.project.ui.actions.support.CommandUtils;
+import org.netbeans.modules.php.project.util.PhpProjectUtils;
+import org.netbeans.modules.php.spi.testing.PhpTestingProvider;
+import org.netbeans.spi.project.SingleMethod;
+import org.openide.filesystems.FileObject;
+import org.openide.util.Lookup;
+import org.openide.util.RequestProcessor;
+
+
+public class ComputeTestMethodAnnotations implements DocumentListener, PropertyChangeListener, Runnable {
+
+    private static final Logger LOGGER = Logger.getLogger(Frameworks.class.getName());
+
+    private static final RequestProcessor RP = new RequestProcessor(ComputeTestMethodAnnotations.class);
+    private final RequestProcessor.Task task = RP.create(this);
+
+    private static final ComputeTestMethodAnnotations INSTANCE = new ComputeTestMethodAnnotations();
+
+    private static boolean isRegistered = false;
+    private static int usagesCount = 0;
+
+    private Document handledDocument;
+
+    public static ComputeTestMethodAnnotations getInstance() {
+        return INSTANCE;
+    }
+
+    private ComputeTestMethodAnnotations() {
+    }
+
+    public synchronized void register() {
+        if (!isRegistered) {
+            EditorRegistry.addPropertyChangeListener(this);
+            isRegistered = true;
+        }
+        usagesCount++;
```

Comment:
e.g. introduce methods?
```java
public static synchronized void incrementUsagesCount() {
    usagesCount++;
}

public static synchronized boolean canRemovePropertyChangeListener() {
    usagesCount--;
    return usagesCount == 0;
}
```

@tmysik Am I wrong?

Reply:
That would work. But in this particular case, using `AtomicInteger` would do the job nicely, without any synchronization (yes, it should rather be `AtomicLong`).